students/DennisLee/lesson06/test_mailroom.py

```diff
@@ -0,0 +1,393 @@
+#!/usr/bin/env python3
+
+import os
+import mailroom as m
+import pytest
+
+
+choices = {  # Dict of command text and functions in the main menu
+'1': {'option': 'Send a thank you', 'function': m.send_thank_you},
+'2': {'option': 'Create a report', 'function': m.create_a_report},
+'3': {'option': 'Send letters to everyone', 'function': m.send_all_letters},
+'4': {'option': 'Quit', 'function': m.exit_screen}
+}
+
+# Dict of functions to show donor list or quit when sending a thank you
+alt_choices = {  
+        '': {'function': m.exit_screen},
+        'quit': {'function': m.exit_screen},
+        'list': {'function': m.print_donor_list}
+}
+
+
+def test_manage_donors_1():
+    # An I/O error occurs because the function prompts for menu choice
+    with pytest.raises(IOError):
+        m.manage_donors()
```

Comment:
Rather than catching these `IOError`s like this for your prompts, I would encourage you to look into the concept of mocking user input data for your tests. Python's `unittest` testing framework uses classes and has a little steeper learning curve than `pytest` but you seem to have a good handle on the basics of unit testing and I think you'd find `unittest` very useful. The `unittest.mock` module allows you to essentially fake user input data, third party API calls etc. in your tests so that you can test the parts of your program that depend on these types of external inputs and resources.

Reply:
Thanks. I had tried reassigning the input name to my own input-mocking function, but that didn't seem to work during testing. Also, the course content is a little confusing with regard to what testing tool to use; it explicitly mentions `pytest`, but then some of the required reading material showed how to use `unittest`. Since `unittest` involved classes, which haven't been covered extensively in the coursework yet, I assumed we were supposed to stick with `pytest`.

Reply:
@DennisLee-DennisLee I totally agree and was already planning on bringing this to the attention of the course authors. We've had students use both `pytest` and `unittest` for the mailroom assignment and it's fine to go with whichever test runner/framework you prefer. Some people like `pytest` for its simplicity and more Pythonic style, but I feel like for both mocks and set up data, you might find the `unittest` API quite useful.